### ISSUE TITLE
Slow _filter_existing_notebooks impacts each submission

### ISSUE THREAD

**Issue by ryanlovett** [jupyter/nbgrader#929]
On a deployment with ~600 students, _filter_existing_notebooks takes about 30s. This hits us when manual grading. A single submission is loaded (/formgrader/submissions/:submission_id) and that invokes api.get_notebook_submission_indices which calls the filter which walks the filesystem and filters out non-existing files. So when the grader presses Next, the next submission is loaded which causes another 30s filesystem walk.

### `nbgrader --version`
0.6.0.dev
### `jupyterhub --version` (if used with JupyterHub)
0.8.1
### `jupyter notebook --version` 
5.4.0
### Expected behavior
Indexes are cached or are built asynchronously?
### Actual behavior
Indexes are gathered for each submission load.
### Steps to reproduce the behavior
Manual grading in big course, visit assignment. Then visit first submission.

**Comment by ryanlovett** [comment]
Re the filter:
> This functionality is necessary for cases where student delete or rename on or more notebooks in their assignment, but still submit the assignment.

Should this invalidate their submission when submitting?

**Comment by ryanlovett** [comment]
I see that one can set ExchangeSubmit.strict=True. So I think _filter_existing_notebooks should be a no-op when this is set?

**Comment by jhamrick** [comment]
Wow, that seems really slow. There probably should be a way to either disable this call if `ExchangeSubmit.strict == True` but I am also curious why it is so slow for you in the first place. I just did a quick test with 600 students and for me it takes less than a second to run `get_notebook_submission_indices` (which is what calls `_filter_existing_notebooks`):

>>>
In [8]: len(api.get_notebook_submission_indices("ps1", "problem1"))
Out[8]: 601

In [9]: %timeit api.get_notebook_submission_indices("ps1", "problem1")
853 ms ± 76 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Then again... this is on my laptop which has a SSD, so I suppose this could be substantially slower on a regular hard drive. What sort of setup do you have for running nbgrader?

**Comment by ryanlovett** [comment]
This is happening from within a Kubernetes pod which attaches volumes via NFS. Our values:
```
len(api.get_notebook_submissions('lab01', 'lab01'))
599

%timeit api.get_notebook_submissions('lab01', 'lab01')
33.7 s ± 989 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by jhamrick** [comment]
Hmm ok, I wonder if it's latency due to NFS that's the problem. That's really surprising that it's so slow; it's approximately 50msec per call to `os.path.exists`.

As a quickfix, I can definitely check whether `ExchangeSubmit.strict == True` and disable calls to `_filter_existing_notebooks` in that case. As a better solution, it might be a good idea to eventually include some sort of caching.

**Comment by jhamrick** [comment]
Closed by #934
